File: algorithms/easy/merge_two_2d_arrays_by_summing_values.dart
class Solution {
  List<List<int>> mergeArrays(List<List<int>> nums1, List<List<int>> nums2) {
    int i = 0;
    int j = 0;
    List<List<int>> res = [];
    while (i < nums1.length && j < nums2.length) {
      if (nums1[i][0] == nums2[j][0]) {
        res.add([nums1[i][0], nums1[i][1] + nums2[j][1]]);
        i++;
        j++;
      } else if (nums1[i][0] < nums2[j][0]) {
        res.add(nums1[i]);
        i++;
      } else {
        res.add(nums2[j]);
        j++;
      }
    }

    while (i < nums1.length) {
      res.add(nums1[i]);
      i++;
    }

    while (j < nums2.length) {
      res.add(nums2[j]);
      j++;
    }

    return res;
  }
}

File: algorithms/easy/merge_two_2d_arrays_by_summing_values.ex
defmodule Solution do
  @spec merge_arrays(nums1 :: [[integer]], nums2 :: [[integer]]) :: [[integer]]
  def merge_arrays([], nums2) do
    nums2
  end

  def merge_arrays(nums1, []) do
    nums1
  end

  def merge_arrays([[id1, v1] | rs1] = nums1, [[id2, v2] | rs2] = nums2) do
    cond do
      id1 == id2 -> [[id1, v1 + v2] | merge_arrays(rs1, rs2)]
      id1 < id2 -> [[id1, v1] | merge_arrays(rs1, nums2)]
      true -> [[id2, v2] | merge_arrays(nums1, rs2)]
    end
  end
end
